### ISSUE TITLE
Plan for Vue CLI v5

### ISSUE THREAD

**Issue by sodatea** [vuejs/vue-cli#6064]
* Development branch: [dev](https://github.com/vuejs/vue-cli/tree/dev), and will be merged to [next](https://github.com/vuejs/vue-cli/tree/next) on each pre-release.
* Documentation: <https://next.cli.vuejs.org/migrations/migrate-from-v4.html>

Here are the rough ideas:

## Major Dependency Upgrades

* [x] Webpack 5 (also allows to opt-out to webpack 4), https://github.com/vuejs/vue-cli/pull/6060
* [x] HTML Webpack Plugin 4
* [x] Cypress 6, would be a peer dependency
* [x] Workbox 6
* [x] Jest 26
* [x] Mocha 8
* [x] ESLint 7
* Other necessary major dependency version bumps, including various webpack loaders and plugins

## Planned New Features

* [ ] Performance
	* Now that we use webpack 5 by default, we can utilize its persistent cache to have a better compilation performance.
* [x] Predicatibility
	* ~~A `--transpile-all` CLI flag to include `node_modules` for transpiration, as a fail-safe for the `transpileDependencies` option.~~ The `transpileDependencies` option can be set to `true` to transpile all dependencies in `node_modules`. It should be the recommended way to build your app, especially when you are using webpack 5, as the compilation overhead should be tolerable with the persistent cache.
* [x] Modern mode
	* Enabled by default, with additional `modern-only` and `no-modern` CLI options
	* A separate `polyfills` chunk by default
	* Better `browserslist` integration. For example, `safari-no-module-fix` can be dropped when `browserslist` config does not target those buggy browsers.
* [x] Configuration
	* Async config loading

## Changed Behaviors

* [x] `vue serve/build` will be an alias of `npm run serve/build`. If an entry is designated and no `package.json` found, it’ll prompt to use `vite` (Vue 3) or `parcel` (Vue 2) for instant prototyping instead.

## Dropped Features

* [x] Drop support of Node.js 8, 11, 13
* [x] Drop support of TSLint
* [x] Drop support of `node-sass`
* [ ] [Drop support of extension-less Vue SFC imports](https://github.com/vitejs/vite/issues/178#issuecomment-630138450)
* [ ] Do not allow creating projects with uppercase names even in Vue CLI UI.

## Other Breaking Changes Under Consideration

* [ ] [Respect XDG specification for the location of configs](https://github.com/vuejs/vue-cli/issues/5649)

**Comment by elevatebart** [comment]
Do we choose to drop extension less SFC imports even for webpack projects or only on vite?

**Comment by sodatea** [comment]
Even for webpack projects. There are edge cases in webpack environment too: https://github.com/vuejs/vue-cli/issues/5307

**Comment by lefreet** [comment]
* `VUE_CLI_SERVICE_CONFIG_PATH` as a cli options ?  muilt `vue.config.js` in  monorepos
* `--debug` mode for server run to output some verbose?  `console` is terrible method to track code of cli
* default show spend time for build process，to optimize.

mark

**Comment by ThreeScreenStudios** [comment]
If possible, please update `postcss-loader` dependency to `4.0.4` from `3.0.0` to add compatibility with Postcss 8/Tailwind 2.0.

**Comment by ModyQyW** [comment]
`eslint-loader` has been deprecated. Why not `eslint-webpack-plugin`?

**Comment by sodatea** [comment]
> * `VUE_CLI_SERVICE_CONFIG_PATH` as a cli options ?  muilt `vue.config.js` in  monorepos
> * `--debug` mode for server run to output some verbose?  `console` is terrible method to track code of cli
> * default show spend time for build process，to optimize.
> 
> mark

These are not major design changes, so can be reserved for later minor releases.
We'll focus on the changes listed in the post first.

The last one is easy, though. I'll take a look if time permits.

**Comment by sodatea** [comment]
> If possible, please update `postcss-loader` dependency to `4.0.4` from `3.0.0` to add compatibility with Postcss 8/Tailwind 2.0.

Yeah, I'm working on that.

**Comment by sodatea** [comment]
> `eslint-loader` has been deprecated. Why not `eslint-webpack-plugin`?

It will be included in the final release.

**Comment by yoyo930021** [comment]
Can we add `jsconfig.json` when create project?
This will work better with Vetur or VueDX.

**Comment by elmatou** [comment]
Can we address #4801  ?
It is really hard to develop & debug Service Worker when we have to rebuild the app on each iteration.

It could mean enabling service worker in development mode, which seems to be a major change.

Thanks

**Comment by fangbinwei** [comment]
> Async config loading

Try to achieve this feature, but block by issue https://github.com/benmosher/eslint-plugin-import/issues/883

sync `@vue/cli-service/webpack.config.js` is required by `@vue/eslint-config-standard`

**Comment by zhouxinyong** [comment]
Can we add `.eslintigonre ` when create project which has eslint?

**Comment by HomyeeKing** [comment]
sometimes it cannot auto import reactive api, like type ref , it can't auto import or using quickfix

**Comment by HomyeeKing** [comment]
when init the project, if choose typescript template, add `ts-jest` automatically also config jest.config

**Comment by sodatea** [comment]
FYI v5.0.0-alpha.0 has been released with most underlying dependencies updated.

We'll work on the planned new features next.
https://github.com/vuejs/vue-cli/releases/tag/v5.0.0-alpha.0

**Comment by 3zzy** [comment]
Is module federation supported for multiple pages / build targets?

**Comment by AGDholo** [comment]
ES feature support in v5: #5881 ?

I created a brand new project using v5 alpha, but it doesn't work ```??``` etc.

**Comment by ChandlerVer5** [comment]
Use the online presets(ajax？or other ways), not just from Git repo! I dont want to create a repo just for a single preset ! 
and if i have two preset in a repo，could I have a choice?

**Comment by Aries-Sushi** [comment]
When is the official version expected to be released?

**Comment by sxzz** [comment]
#2138 
Will we consider supporting vue.config.ts?

**Comment by dreamfinally** [comment]
When is the official version expected to be released?

**Comment by RaafatTurki** [comment]
respecting the XDG specification shouldn't be listed as a breaking change, vue could fallback to legacy paths if the XDG ones (or their fallbacks)

**Comment by RaafatTurki** [comment]
Respecting the XDG specification shouldn't be listed as a breaking change, Vue could always fallback to legacy paths if the XDG ones (or their fallbacks) fail, I've just updated my issue #5649 regarding the matter.

**Comment by Mister-Hope** [comment]
Hi @sodatea , can the alpha version work now?

I just created a new project using `5.0.0-alpha.3`, and I can not run the dev server:

```
95% emitting emit(node:17872) [DEP_WEBPACK_COMPILATION_ASSETS] DeprecationWarning: Compilation.assets will be frozen in future, all modifications are deprecated.
BREAKING CHANGE: No more changes should happen to Compilation.assets after sealing the Compilation.
        Do changes to assets earlier, e. g. in Compilation.hooks.processAssets.
        Make sure to select an appropriate stage from Compilation.PROCESS_ASSETS_STAGE_*.
(Use `node --trace-deprecation ...` to show where the warning was created)
99% done plugins FriendlyErrorsWebpackPlugin

 ERROR  Failed to compile with 1 error                                                                                                   上午11:55:55
  Error: The loader "F:\template\vue\node_modules\html-webpack-plugin\lib\loader.js!F:\template\vue\public\index.html" didn't return html.

  - index.js:336 HtmlWebpackPlugin.evaluateCompilationResult
    [vue]/[html-webpack-plugin]/index.js:336:24

  - index.js:243
    [vue]/[html-webpack-plugin]/index.js:243:22

  - runMicrotasks

  - task_queues.js:93 processTicksAndRejections
    internal/process/task_queues.js:93:5

  - async Promise.all

  - async Promise.all


No issues found.
```

**Comment by cexbrayat** [comment]
The error comes from an incompatibility between webpack 5.22.0 (that was just released) and the `html-webpack-plugin` v4.5.1. 

This can be fixed by setting your webpack version to `5.21.2` for now in your package.json

The proper fix is to update the html-webpack plugin to v5.1.0 (see https://github.com/jantimon/html-webpack-plugin/issues/1603) . This has already been done by @sodatea in Vue CLI (see https://github.com/vuejs/vue-cli/commit/4ce7edd3754c3856c760d126f7fa3928f120aa2e), and will be out in the next release!

**Comment by Ge-yuan-jun** [comment]
> Is module federation supported for multiple pages / build targets?

module federation is a feature of webpack5, and when I use it in pure webpack without any cli, it works, 

here is the link: https://github.com/Ge-yuan-jun/pure-webpack5-mf-plugin

**Comment by Ge-yuan-jun** [comment]
Have a issue with v5.0.0-alpha.5

#6318

**Comment by Snapu** [comment]
In webpack 5 you can set output.publicPath: "auto". Which is a great functionality in combination with module federation. Would be great if vue cli option "publicPath" would support it too as modifying it directly is not allowed.

**Comment by trainoasis** [comment]
Was wondering, probably Cypress should be v6 now, not v5? Or is it intentional?

**Comment by panhezeng** [comment]
https://github.com/neutrinojs/webpack-chain/issues/266 webpack-chain alternative plan ？

**Comment by T0miii** [comment]
@Ge-yuan-jun did you manage to run module federation with vue cli 5.0.0.8 alpha ?

**Comment by sodatea** [comment]
> Was wondering, probably Cypress should be v6 now, not v5? Or is it intentional?

The latest alpha version uses v6 now. It was v5 when the issue was opened.

> [neutrinojs/webpack-chain#266](https://github.com/neutrinojs/webpack-chain/issues/266) webpack-chain alternative plan ？

`webpack-chain` is still gonna be maintained. So no need to find an alternative.

**Comment by falstack** [comment]
add `SSG` support

**Comment by corbinu** [comment]
@falstack projects like https://nuxtjs.org handle SSG. Just my feeling but it seems out of scope for the CLI

**Comment by nickolboy** [comment]
有没有个大概的时间

**Comment by sodatea** [comment]
My updated plan:

Blocking issues for Beta:
- Changes related to modern mode. Working on that. May be out next week.
- Everything else can either be added or fixed later, as they are not likely to be breaking changes.

Things must be done before RC:
- (Blocking) A good enough default cache configuration, so that the build speed is tolerable. At the moment, there are 2 PRs working on that https://github.com/vuejs/vue-cli/pull/6323 https://github.com/vuejs/vue-cli/pull/6390 but we need more testing during the betas.

Before stable release:
- Mochapack webpack 5 support. https://github.com/sysgears/mochapack/pull/83 But if that's too hard, I think I'll go with webpack 4 anyway. And I believe the planned webpack 5 support could be backward compatible so it's not a blocking issue.

**Comment by Akryum** [comment]
@sodatea Did you check the [lazy compilation](https://webpack.js.org/configuration/experiments/#experimentslazycompilation) mode for Webpack 5 out? Could massively increase performance for big apps

**Comment by sodatea** [comment]
> @sodatea Did you check the [lazy compilation](https://webpack.js.org/configuration/experiments/#experimentslazycompilation) mode for Webpack 5 out? Could massively increase performance for big apps

Yeah, it's marked as experimental, so I haven't decided whether to enable it by default or leave it as-is.

**Comment by TrungRueta** [comment]
Hi @sodatea 
I have just install global `@vue/cli@next` and do upgrade project (from vue cli v4) follow instruction in next.vue-cli site. I see a issue: when setup everything done, run `vue-cli-service serve`, it running thought with message:
```
 INFO  Starting development server...
99% done plugins FriendlyErrorsWebpackPlugin

 DONE  Compiled successfully in 57361ms                                                                                                     6:59:16 PM

99% done plugins vue-cli-service serve
  App running at:
  - Local:   http://localhost:8080/
  - Network: http://localhost:8080/

  Note that the development build is not optimized.
  To create a production build, run npm run build.

99% done plugins FriendlyErrorsWebpackPlugin

 DONE  Compiled successfully in 2336ms                                                                                                      6:59:20 PM

99% done plugins vue-cli-service serve
  App running at:
  - Local:   http://localhost:8080/
  - Network: http://localhost:8080/
 ```

Issue "99% ..." message showing from Alpha versions to until now. Site still work ok, but message a bit confuse, no see any error come next. 

Other issue is when press cmd + c to terminate progress , in v4 it immediate disrupt dev server but in v5 it hangup few sec (amount ~ 4 5 secs), must press several times before it complete disrupt it show:

```
99% end closing watch compilation%
```

Can you help validate this issue ? i can not find any topic say about this.

**Comment by sodatea** [comment]
@TrungRueta

The `99%` issue is caused by webpack 5 continuing sending progress events after compilation is done (for things like caching).
It doesn't affect the actual compilation progress.

I've been trying to fix the output message. Any help is also appreciated.

---

Didn't know about the Ctrl + C issue before. I'll have a look.

**Comment by TrungRueta** [comment]
I have not much experience about deep package scripting, but i will try to help clarify reason. I tested with fresh new project compare with my current project.

Fresh new project created from `vue create <>` is lightweight and only delay very short time when terminate. clip when start and stop it here

https://user-images.githubusercontent.com/12226007/116001293-f46d5380-a61d-11eb-8347-3f117fb1f309.mov

this is setup:
![Screen Shot 2021-04-25 at 10 57 29 PM](https://user-images.githubusercontent.com/12226007/116001304-02bb6f80-a61e-11eb-9a28-f8144ccc1f8e.png)

you can see 99% message, note that dev server only show process one time (finish after first message 99% shown). And when terminate dev server (you can see syntax `^c` it is pressing `ctrl + c`) it delay 0.5s.

---

in second i running v5 on my current project which real and heavier. this is result:


https://user-images.githubusercontent.com/12226007/116001344-3dbda300-a61e-11eb-945c-c95fb17abf8b.mov

This time it show 99% at near end or project but then trigger compile progress again (or maybe bug message only, i dont know yet). Result is we have 2 complete notification. Later when press `ctrl + c` it give hug delay , i tried press multi times buttons to demo that in delay time can do nothing. 

this is packages.json of my project. 

```json
{
"dependencies": {
    "@apollo/client": "^3.3.11",
    "@aws-amplify/auth": "^3.4.26",
    "@aws-amplify/core": "^3.8.18",
    "@fortawesome/fontawesome-svg-core": "^1.2.34",
    "@fortawesome/free-brands-svg-icons": "^5.15.2",
    "@fortawesome/pro-duotone-svg-icons": "^5.15.3",
    "@fortawesome/pro-light-svg-icons": "^5.15.2",
    "@fortawesome/pro-solid-svg-icons": "^5.15.2",
    "@fortawesome/vue-fontawesome": "^2.0.2",
    "@vue/composition-api": "^1.0.0-rc.2",
    "buefy": "^0.9.5",
    "chart.js": "^2.9.4",
    "comlink": "^4.3.0",
    "core-js": "^3.8.1",
    "d3": "^6.5.0",
    "faker": "^5.4.0",
    "fuse.js": "^6.4.6",
    "graphql": "^15.5.0",
    "luxon": "^1.26.0",
    "md5": "^2.3.0",
    "numeral": "^2.0.6",
    "portal-vue": "^2.1.7",
    "register-service-worker": "^1.7.2",
    "rxjs": "^6.6.3",
    "socket.io-client": "^4.0.0",
    "subscriptions-transport-ws": "^0.9.18",
    "uuid": "^8.3.2",
    "v-calendar": "^2.2.4",
    "vue": "^2.6.11",
    "vue-chartjs": "^3.5.1",
    "vue-emoji-picker": "^1.0.1",
    "vue-router": "^3.4.3",
    "vue-select": "^3.11.2",
    "vue-toastification": "^1.7.11",
    "yup": "^0.32.9"
  },
  "devDependencies": {
    "@types/d3": "^6.3.0",
    "@types/faker": "^5.1.7",
    "@types/jest": "^26.0.22",
    "@types/luxon": "^1.26.0",
    "@types/numeral": "^2.0.0",
    "@types/uuid": "^8.3.0",
    "@types/yup": "^0.29.11",
    "@vue/cli-plugin-babel": "~5.0.0-beta.0",
    "@vue/cli-plugin-pwa": "~5.0.0-beta.0",
    "@vue/cli-plugin-router": "~5.0.0-beta.0",
    "@vue/cli-plugin-typescript": "~5.0.0-beta.0",
    "@vue/cli-plugin-unit-jest": "~5.0.0-beta.0",
    "@vue/cli-service": "~5.0.0-beta.0",
    "@vue/runtime-dom": "^3.0.11",
    "@vue/test-utils": "^1.1.3",
    "@vuedx/typescript-plugin-vue": "^0.6.3",
    "prettier": "^2.2.1",
    "sass": "^1.30.0",
    "sass-loader": "^10.1.0",
    "typescript": "^4.2.3",
    "vue-template-compiler": "^2.6.11",
    "windicss": "^2.5.9",
    "windicss-webpack-plugin": "^0.3.2",
    "worker-loader": "^3.0.8"
  }
}
```

this is vue.config:

```js
const WebpackWindiCSSPlugin = require('windicss-webpack-plugin').default;

module.exports = {
  pwa: {
    workboxOptions: {
      skipWaiting: true,
      clientsClaim: true,
    },
  },
  devServer: {
    host: 'localhost',
  },
  chainWebpack: (config) => {
    config.plugins.delete('fork-ts-checker').end();

    config.module
      .rule('worker')
      .before('ts')
      .test(/\.worker\.ts$/i)

      .use('worker-loader')
      .loader('worker-loader')
      .options({
        filename: '[name].[fullhash:8].js',
        // inline: 'fallback',
        worker: {
          type: 'SharedWorker',
          options: {
            credentials: 'omit',
            name: 'GrowApp', // for debbuging
          },
        },
      })
      .end();
  },
  configureWebpack: {
    plugins: [
      new WebpackWindiCSSPlugin({
        config: `tailwind.config.js`,
        scan: {
          dirs: ['./src'],
          exclude: ['node_modules', '.git', 'public/**/*', '*.template.html', 'index.html', 'dist'],
          include: [],
        },
        transformCSS: 'pre',
        compile: false,
      }),
    ],
  },
};
```

Hope there info can help team find issue!

**Comment by WilliamALiang** [comment]
When I package, the files in the public directory will also be compressed? Packing time is much longer than before.

version: 5.0.0-beta.1

**Comment by bodograumann** [comment]
Maybe you already know about this, but I just wanted to mention the remaining dependency conflicts (with npm 7) that I encountered when migrating to vue-cli 5.0.0-beta.2:

> npm WARN Could not resolve dependency:       
> npm WARN peer vue@"^2.x" from vue-jest@4.0.1
> npm WARN node_modules/vue-jest              
> npm WARN   dev vue-jest@"^4.0.1" from the root project
> npm WARN   1 more (@vue/cli-plugin-unit-jest)
>
> npm WARN Could not resolve dependency:
> npm WARN peer webpack@"^4.0.0" from cache-loader@4.1.0
> npm WARN node_modules/cache-loader
> npm WARN   cache-loader@"^4.1.0" from @vue/cli-service@5.0.0-beta.2
> npm WARN   node_modules/@vue/cli-service
> npm WARN   1 more (@vue/cli-plugin-typescript)

The first one comes from vue 3 being installed. I guess in that case vue-jest 5 is needed.

Apart from this minor problem the upgrade went quite smoothly. So thank you for that :heart:

**Comment by jfbrennan** [comment]
Can we just make Vue SFC the only way, optimize a single compilation step that, and do away with literally everything else? I have spent days slogging through build crap and I like to think my project is pretty slimmed down and simple (I don't have TS, I don't have SASS or any other loaders, and it's all still a major PIA!)

**Comment by a951055** [comment]
I have a proposal, use 'glob' and 'require.resolve ' scan `@vue/cli-plugin-*`, replace the old way of querying through this.pkg.
example:
```
const projectPlugins = globs
    .sync([
        './node_modules/@vue/cli-plugin-*',
        './node_modules/vue-cli-plugin-*',
        './node_modules/@**/vue-cli-plugin-*'
    ])
    .map((depPath) => depPath.replace(/\.\/node_modules\//, ''))
    .filter(isPlugin)
    .map((id) => {
        if (this.pkg.optionalDependencies && id in this.pkg.optionalDependencies) {
            let apply = loadModule(id, this.pkgContext)
            if (!apply) {
                warn(`Optional dependency ${id} is not installed.`)
                apply = () => {}
            }

            return { id, apply }
        } else {
            return idToPlugin(id, resolveModule(id, this.pkgContext))
        }
    })


```
// link: https://github.com/vuejs/vue-cli/blob/882aaca1fc80f7f478bac2f309b5e9b7f1acfe6b/packages/%40vue/cli-service/lib/Service.js#L182
old:
```
 const projectPlugins = Object.keys(this.pkg.devDependencies || {})
                .concat(Object.keys(this.pkg.dependencies || {}))
                .filter(isPlugin)
                .map((id) => {
                    if (this.pkg.optionalDependencies && id in this.pkg.optionalDependencies) {
                        let apply = loadModule(id, this.pkgContext)
                        if (!apply) {
                            warn(`Optional dependency ${id} is not installed.`)
                            apply = () => {}
                        }

                        return { id, apply }
                    } else {
                        return idToPlugin(id, resolveModule(id, this.pkgContext))
                    }
                })

```
If use this, we can use `base config` configure project

**Comment by DanSnow** [comment]
@a951055 This will break [Yarn PnP](https://yarnpkg.com/features/pnp).

**Comment by livthomas** [comment]
I have noticed two issues when running `vue upgrade --next` in my project:

1. `yarn` was used even if I only use `npm`. I guess this is probably because I ran this command in a subfolder of an npm monorepo where there is no `package-lock.json` file.
2.  `@vue/cli-*` packages were added to `dependencies` instead of `devDependencies` where the old 4.x packages were left untouched.

**Comment by a951055** [comment]
@DanSnow I don't think so. because. I think that through shallow scanning is not much (scan level 1 and level 2 in '/node_modules'). and can support your project use BaseConfig, such as maven. 
This is very useful for a large number of projects. that's what babel did.

**Comment by DanSnow** [comment]
@a951055 Yarn PnP doesn't use `node_modules`. Instead, it saves all of the packages in a flat structure and patch the `require` to resolve a package name to correct path. Here is how a project with Yarn PnP will look like:
![image](https://user-images.githubusercontent.com/5575082/124077023-932c8b00-da79-11eb-9847-067485a30e6d.png)

**Comment by a951055** [comment]
@DanSnow i see. I understand wrong.

**Comment by a951055** [comment]
I'm here again. This time I want to change my ideas. I want to change it into "Vue plugin" by interface loading`. 
I need this interface

**Comment by bodograumann** [comment]
@a951055 Please don’t spam this thread. You can create a feature request. Feel free to also provide a PR with the implementation while you are at it.

**Comment by jacekkarczmarczyk** [comment]
are there plans to migrate to `webpack-dev-server@4` (currently in beta)?

**Comment by IndexXuan** [comment]
I am the author of [vue-cli-plugin-vite](https://github.com/IndexXuan/vue-cli-plugin-vite).
As of version v1.3.5, basic webpack config and the different of webpack ecosystem and vite is compat(e.g. env vars, css-loader, devServer.{before, public, host} and so on). We can also use html template in vite(public/index.html). 

Problems we cannot campatible is the **Heavy use** chainWebpack or **Heavy vue-cli-plugin** like `electron plugin /  ssr plugin`.
for heavy use chainWebpack, we can warning users writing **universal vue.config.js**, like check plugin/args exists before modify/tap it.
```js
// vue.config.js
if (config.plugins.has('xxx') {
  // then modify it. because in vue-cli-plugin-vite mode,
  // there are no real webpack-plugin like html-webpack-plugin.
}
```

Benefits, we use **real vite** in exists vue-cli project and migrate to vite in the future is smoothly. The **BEST DX** today with lighting fast devServer start and HMR.
PR for [vue-enterprise-boilerplate](https://github.com/IndexXuan/vue-enterprise-boilerplate/compare/master...IndexXuan:feature/vite), less than 20 lines codebase modifications and we successfully use vite in vue-enterprise-boilerplate. or you can vue create my-vite-app && vue add vite.

Do you think we can (or possible) use vite for development and webpack(vue-cli-service build) for production ?
What is the **Future Planning** of vue-cli and vite ?

Thanks.

**Comment by akkayy** [comment]
![image](https://user-images.githubusercontent.com/73772701/126434716-2a702893-5d2c-4691-a179-908795375a92.png)

I raised this with vite and see Evan's reply here - https://github.com/vitejs/vite/issues/1527 

Please clarify how is Instant Prototyping possible with Vite.

**Comment by zdm** [comment]
Do we have any real terms when v5 will be released?

**Comment by nVitius** [comment]
Is anyone working on upgrading ts-loader to v9 on the typescript plugin?

**Comment by zdm** [comment]
v5 version receive almost no commits.
maybe its time to release it and add improvements later?

**Comment by ffxsam** [comment]
Do you think it would be feasible to drop webpack completely, and go with esbuild instead? I've been using it lately and it's _incredibly_ fast, way less bloat for sure.

**Comment by alireza4050** [comment]
Is webpack-dev-server v4 going to be supported? It's [rc0](https://github.com/webpack/webpack-dev-server/releases/tag/v4.0.0-rc.0) version has been released for a month.

**Comment by nVitius** [comment]
I created a PR for removing cache-loader in favor of webpack's own filesystem cache (webpack 5).
https://github.com/vuejs/vue-cli/pull/6634

**Comment by sgielen** [comment]
#6669 seems to be the PR for updating to webpack-dev-server v4, so it's in progress :-)

**Comment by ffxsam** [comment]
@IndexXuan @bodograumann I'm curious about the 👎 to my esbuild suggestion, since I didn't see any written-out responses. Is there a reason you feel esbuild would be a bad direction?

**Comment by DRoet** [comment]
@ffxsam I don't think it is realistic to drop webpack from `@vue/cli` as it would break the entire ecosystem around it (plugins/loaders). If you want to use something more modern, you should probably look into [vite](https://vitejs.dev/) since that already uses esbuild for some parts of it.

**Comment by ffxsam** [comment]
@DRoet Thanks - that makes total sense.

**Comment by Mister-Hope** [comment]
```
options has an unknown property 'overlay'.
```

Got something imcompatable with latest `webpack-dev-server` and `beta.4`

Also, I am not sure, but jest plugin seems imcompatable with latest jest:

```
 FAIL  tests/unit/example.spec.ts
  ● Test suite failed to run

    TypeError: Cannot destructure property 'config' of 'undefined' as it is undefined.

      at Object.getCacheKey (node_modules/vue-jest/lib/index.js:10:7)
      at ScriptTransformer._getCacheKey (node_modules/@jest/transform/build/ScriptTransformer.js:280:41)
      at ScriptTransformer._getFileCachePath (node_modules/@jest/transform/build/ScriptTransformer.js:351:27)     
      at ScriptTransformer.transformSource (node_modules/@jest/transform/build/ScriptTransformer.js:588:32)       
      at ScriptTransformer._transformAndBuildScript (node_modules/@jest/transform/build/ScriptTransformer.js:758:40)
      at ScriptTransformer.transform (node_modules/@jest/transform/build/ScriptTransformer.js:815:19)

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        14.588 s
Ran all test suites.
```

**Comment by cexbrayat** [comment]
@Mister-Hope If you want to use Jest v27, you need to use the new package `@vue/vue3-jest` (or `@vue/vue2-jest`). Jest v27 had breaking changes, so we had to release a new version of `vue-jest`.

See https://github.com/vuejs/vue-cli/pull/6627 for more info on how to configure your project (or create a new project to check the configuration). If you run into an issue, feel free to open a dedicated issue and ping me!

**Comment by Mister-Hope** [comment]
> @Mister-Hope If you want to use Jest v27, you need to use the new package `@vue/vue3-jest` (or `@vue/vue2-jest`). Jest v27 had breaking changes, so we had to release a new version of `vue-jest`.
> 
> See #6627 for more info on how to configure your project (or create a new project to check the configuration). If you run into an issue, feel free to open a dedicated issue and ping me!

Thanks for that reminder, I deleted `vue-jest`, and changed my `jest.config,js`, then unit test works fine now.

@sodatea I think we should better mark all the v27 alpha versions of `vue-jest` as deprecated to warn the users like me or especially mention this in changelog. Since V5 has been devloped for 9 month, I think there will be plenty of man switching to v5 alpha versions with webpack 4 to get full power of node 12 or 14. And this deps is manually decalred in package.json. instead of on the tool chain.

**Comment by lloydjatkinson** [comment]
Is there any timeline on this?

**Comment by ahnpnl** [comment]
Hi, is it possible to make Jest as peer dep? Vue CLI 4 doesn’t allow overwrite Jest dep (I haven’t tested Vue CLI 5 RC yet)

**Comment by BenJackGill** [comment]
Evan has [gone on the record](https://github.com/vuejs/rfcs/discussions/378#discussioncomment-1197332) saying `<script setup>` is intended to be the standard syntax for Vue single-file components going forward.

But right now the Vue Cli scaffolds single-file components using `export default defineComponent({ etc etc });`.

Therefore I suggest using `<script setup>` as the default scaffolding when selecting Vue 3. At the very least it should be an option during project creation. This will dramatically help with uptake, and move the community towards the intended standard syntax as outlined by Evan.

**Comment by akkayy** [comment]
> Evan has [gone on record](https://github.com/vuejs/rfcs/discussions/378#discussioncomment-1197332)

He also went on record on [reddit](https://www.reddit.com/r/vuejs/comments/r1vluc/new_default_recommendations/?utm_source=share&utm_medium=mweb3x) & [twitter](https://twitter.com/youyuxi/status/1463429442076745730?t=n1RJ5BydUGOr5rS_fA8iWA)... 

These pictures summarizes it:

![1D62DDBA-C1CC-444D-9CE1-AA90215BF5FA](https://user-images.githubusercontent.com/73772701/144972405-e8157e4b-d970-42e9-ba96-c671437e3c57.jpeg)

![05D0E138-D6E1-48B2-9186-D3E91E6BF8D6](https://user-images.githubusercontent.com/73772701/144972492-7ebb0678-2990-4936-b698-de3711e7d3a3.jpeg)

... makes us wonder what the future of Vue-CLI is?

**Comment by sodatea** [comment]
> Hi, is it possible to make Jest as peer dep? Vue CLI 4 doesn’t allow overwrite Jest dep (I haven’t tested Vue CLI 5 RC yet)

Then we need separate packages for Jest presets. Ideally two, one for Vue 2 and one for Vue 3.
Then the complexity may grow out of control.
So at the moment, I prefer to keep it inside the plugin.
And we may bump Jest major versions more often after v5 is out.

**Comment by sodatea** [comment]
> Therefore I suggest switching to `<script setup>` as the default scaffolding when selecting Vue 3. At the very least it should be an option during project creation.

I'd love to but I don't have the time. So PR's welcome.
Actually, the new landing page in `create-vue` was originally implemented for Vue CLI.
But it messed up all the CLI tests and I can't fix them in a reasonable time frame. So it ends up in that new project without any tech debts. The CLI templates are remained as-is.

**Comment by sodatea** [comment]
> ... makes us wonder what the future of Vue-CLI is?

Vue CLI will continue to be maintained and still the recommended tooling for webpack users.
Webpack is not going to fade away any time soon so Vue CLI isn't gonna be abandoned either.

But we must say, Vite and `create-vue` provide a better out-of-box experience for Vue users to start a new project.
That's why they are the "default recommendations".

**Comment by ahnpnl** [comment]
> > Hi, is it possible to make Jest as peer dep? Vue CLI 4 doesn’t allow overwrite Jest dep (I haven’t tested Vue CLI 5 RC yet)
> 
> Then we need separate packages for Jest presets. Ideally two, one for Vue 2 and one for Vue 3. Then the complexity may grow out of control. So at the moment, I prefer to keep it inside the plugin. And we may bump Jest major versions more often after v5 is out.

I think a warning or error to ask users to install Jest as a dependency when trying to use the preset package could be an idea? 

I think the idea to let CLI to install Jest when generating the project is good, but it shouldn’t prevent the users from upgrading Jest themselves whenever they need.

**Comment by cexbrayat** [comment]
@ahnpnl We made Jest an explicit dep in CLI v5, you can give it a try (see https://github.com/vuejs/vue-cli/pull/6627). A new project now uses Jest v27 and `@vue/vue3-jest` instead of `vue-jest`

**Comment by yoyo837** [comment]
Can you provide an optional configuration of `core-js@2`?

**Comment by ahnpnl** [comment]
> @ahnpnl We made Jest an explicit dep in CLI v5, you can give it a try (see #6627). A new project now uses Jest v27 and `@vue/vue3-jest` instead of `vue-jest`

Tested out with CLI v5 it works fine. I only can test it my personal project, my work project won't allow to use CLI v5 until an official release is out.

Another idea is: since Vue 3 drops IE11 support IIRC, can we provide an option to opt-in/out `core-js`?

**Comment by lorand-horvath** [comment]
@sodatea May I ask when are we to expect the release of the final version of Vue CLI 5 ?

**Comment by lorand-horvath** [comment]
> @ahnpnl Another idea is: since Vue 3 drops IE11 support IIRC, can we provide an option to opt-in/out `core-js`?

Perhaps keeping it simple would be a better idea: `core-js` used by default for Vue 2 projects (it still needs to support IE), but removed in Vue 3 projects (completely drop IE support)?

**Comment by ahnpnl** [comment]
> > @ahnpnl Another idea is: since Vue 3 drops IE11 support IIRC, can we provide an option to opt-in/out `core-js`?
> 
> Perhaps keeping it simple would be a better idea: `core-js` used by default for Vue 2 projects (it still needs to support IE), but removed in Vue 3 projects (completely drop IE support)? Don't know if this is really feasible, though.

This option LGTM

**Comment by darrinmn9** [comment]
@ahnpnl just want to clarify core-js usage since there isnt much documentation on it. If I have a vue2 app but only need to support modern/esm browsers (not ie11), is core-js doing anything of value or could i just remove it?

**Comment by ahnpnl** [comment]
Hi, in short, you only need `core-js` if you need to support ie11. It's not specific to vue2 or vue3.

**Comment by jacekkarczmarczyk** [comment]
> Hi, in short, you only need core-js if you need to support ie11

I don't think it's true, `core-js` provides various polyfills for features that are not available in other browsers too, for example `Array.findLast` which is not available in current Chrome and many more

**Comment by lorand-horvath** [comment]
> > Hi, in short, you only need core-js if you need to support ie11
> 
> I don't think it's true, `core-js` provides various polyfills for features that are not available in other browsers too, for example `Array.findLast` which is not available in current Chrome and many more

Array's findLast() method is in fact available in latest Chrome 97 https://caniuse.com/?search=findLast
But is Vue 2/3 or Vue CLI actually using it anywhere? I very much doubt it...

**Comment by yoyo837** [comment]
> Array's findLast() method is in fact available in latest Chrome 97 https://caniuse.com/?search=findLast
> But is Vue 2/3 or Vue CLI actually using it anywhere? I very much doubt it...

Yes, you are right.

Maybe someone else has some misunderstanding of what `core-js` does.

**Comment by petterw03** [comment]
What happens to v5 if this project is now in maintenance mode? Or is `create-vue` the actual stable release? This is very confusing.

**Comment by lloydjatkinson** [comment]
> What happens to v5 if this project is now in maintenance mode? Or is `create-vue` the actual stable release? This is very confusing. 

The entire official stance on any Vue tooling has been disjointed and a mess ever since Vue 3.

**Comment by sodatea** [comment]
1. I'll continue maintaining this repo, applying patches, adding small new features.
2. I'll make no major changes to this project.
3. We recommend new projects to use `create-vue`.

---

Regarding the v5 stable release:

1. In most cases you can just use the RC version, it's quite stable IMO;
2. I would very much like to ship a new stable release, too. But I'm not confident enough to do so because I'm not ready for the incoming questions and issues once it's out.
3. Currently, my plan is to check again and do some last-minute bug fixes and ship another RC then stable.

**Comment by bodograumann** [comment]
I think one important thing that is still missing from v5.rc is the typechecking of script-setup sfcs.

**Comment by sodatea** [comment]
> I think one important thing that is still missing from v5.rc is the typechecking of script-setup sfcs.

If `fork-ts-checker-webpack-plugin` still can't fully support that, we may end up recommending using `vue-tsc` separately for type-checking.

**Comment by lorand-horvath** [comment]
@sodatea You said new projects should use `create-vue`. OK, but for us, who are used to vue-cli (I've been using it for a long time), there's also `vite`... so which one should we use. The ecosystem is becoming fragmented, without any obvious explanation as to why. Isn't vue spreading itself too thin? I don't blame anyone, just asking... it doesn't seem to go in the right direction IMHO.

**Comment by sodatea** [comment]
`create-vue` is a scaffolding tool. It creates a Vue project that uses Vite for dev server and bundling.
Vite has about 400k/week downloads. It's mature enough for a new project.

You can keep using Vue CLI if you prefer webpack. It's still being maintained.

But Vite and `create-vue` provides a better out-of-box experience.

**Comment by lorand-horvath** [comment]
@sodatea Thanks for the explanation. For anyone out there still wondering what is going on and why there is no documentation, well... here's the new documentation at staging.vuejs.org which is currently kind of "hidden" and which will most probably become the official vuejs.org
https://staging.vuejs.org/guide/scaling-up/tooling.html#project-scaffolding

**Comment by lorand-horvath** [comment]
>     Currently, my plan is to check again and do some last-minute bug fixes and ship another RC then stable.

@sodatea Waiting patiently for the next Vue CLI 5 RC / stable version... I'm not ready to switch to Vite/create-vue for production yet.

**Comment by lpj145** [comment]
For anyone feels missed with vue-ecosystem please, keep in mind what now vue is more easily to develop and deploy than vue 1 is launched, so to a bunch of reasons: vite is more simplest, lazy, fast way to develop, deploy and delivery like vue-cli do in the past, IMHO the vite not need a large list to reasoning with devs about why adopt vite over the vue-cli, reading the vite documentation is enough to understand the fact: now the things is more simple than the past, so trust in vue-core, we understand a lot what vue community need to improves our jobs ❤️

**Comment by lorand-horvath** [comment]
@lpj145 Dude... I read that twice... now my brain hurts.

**Comment by lpj145** [comment]
@lorand-horvath i don't know why, if my words feels aggressive don't worry, so if the fact of vite is now the boss, for me is like train passing above me, but it's life 🤣🤣

**Comment by DRoet** [comment]
According to the last [RC](https://github.com/vuejs/vue-cli/releases/tag/v5.0.0-rc.3). v5 will be stable next week if there are no major regressions found.

**Comment by jfuehner** [comment]
@lorand-horvath, I would love to switch to vite if it had support for async web components. I opened a [discussion thread](https://github.com/vitejs/vite/discussions/2297) on the vite repo a year ago requesting the feature.

**Comment by haoqunjiang** [comment]
v5.0.1 released.
Sorry that I've burnt out for quite some time last year and failed to do a stable release earlier.

There are still a few features that I wish I could add to this release but haven't finished (e.g. creating new projects with ESLint 8).
I plan to continue working on them for the following minor releases.

But anyway, this new major version is stable and ready for use. Please enjoy.

**Comment by alimony** [comment]
@sodatea Thank you for all your hard work on the project.

**Comment by lorand-horvath** [comment]
@sodatea **_THANK YOU_**

**Comment by AnWeber** [comment]
@sodatea thx

**Comment by Keavon** [comment]
Thank you so much @sodatea, it's always stressful maintaining an open source project so many people rely on and it so often goes unappreciated. But speaking on behalf of the thousands of people who haven't even stumbled on this thread, thank you for helping improve our lives as developers!

**Comment by VindyaKaruturi** [comment]
@sodatea  we have updated the vue-cli version to 5.0.1 version for an existing project. This throws below error.

ValidationError: Invalid options object. Dev Server has been initialized using an options object that does not match the API schema.
         - options has an unknown property 'overlay'. These properties are valid:
           object { allowedHosts?, bonjour?, client?, compress?, devMiddleware?, headers?, historyApiFallback?, host?, hot?, http2?, https?, ipc?, liveReload?, magicHtml?, onAfterSetupMiddleware?, onBeforeSetupMiddleware?, onListening?, open?, port?, proxy?, server?, setupExitSignals?, setupMiddlewares?, static?, watchFiles?, webSocketServer? }
ValidationError: Invalid options object. Dev Server has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'overlay'. These properties are valid:
   object { allowedHosts?, bonjour?, client?, compress?, devMiddleware?, headers?, historyApiFallback?, host?, hot?, http2?, https?, ipc?, liveReload?, magicHtml?, onAfterSetupMiddleware?, onBeforeSetupMiddleware?, onListening?, open?, port?, proxy?, server?, setupExitSignals?, setupMiddlewares?, static?, watchFiles?, webSocketServer? }
    at validate (C:\Users\o885500\Vista\azj-geolocation-ui\node_modules\webpack-dev-server\node_modules\schema-utils\dist\validate.js:115:11)
    at new Server (C:\Users\o885500\Vista\azj-geolocation-ui\node_modules\webpack-dev-server\lib\Server.js:232:5)
    at serve (C:\Users\o885500\Vista\azj-geolocation-ui\node_modules\@vue\cli-service\lib\commands\serve.js:183:20)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
error Command failed with exit code 1.

**Comment by yoyo837** [comment]
> @sodatea we have updated the vue-cli version to 5.0.1 version for an existing project. This throws below error.
> 
> ValidationError: Invalid options object. Dev Server has been initialized using an options object that does not match the API schema. - options has an unknown property 'overlay'. These properties are valid: object { allowedHosts?, bonjour?, client?, compress?, devMiddleware?, headers?, historyApiFallback?, host?, hot?, http2?, https?, ipc?, liveReload?, magicHtml?, onAfterSetupMiddleware?, onBeforeSetupMiddleware?, onListening?, open?, port?, proxy?, server?, setupExitSignals?, setupMiddlewares?, static?, watchFiles?, webSocketServer? } ValidationError: Invalid options object. Dev Server has been initialized using an options object that does not match the API schema.
> 
> * options has an unknown property 'overlay'. These properties are valid:
>   object { allowedHosts?, bonjour?, client?, compress?, devMiddleware?, headers?, historyApiFallback?, host?, hot?, http2?, https?, ipc?, liveReload?, magicHtml?, onAfterSetupMiddleware?, onBeforeSetupMiddleware?, onListening?, open?, port?, proxy?, server?, setupExitSignals?, setupMiddlewares?, static?, watchFiles?, webSocketServer? }
>   at validate (C:\Users\o885500\Vista\azj-geolocation-ui\node_modules\webpack-dev-server\node_modules\schema-utils\dist\validate.js:115:11)
>   at new Server (C:\Users\o885500\Vista\azj-geolocation-ui\node_modules\webpack-dev-server\lib\Server.js:232:5)
>   at serve (C:\Users\o885500\Vista\azj-geolocation-ui\node_modules@vue\cli-service\lib\commands\serve.js:183:20)
>   at processTicksAndRejections (node:internal/process/task_queues:96:5)
>   error Command failed with exit code 1.


`webpack-dev-server` modified their api. You need to check their docs for break changes.

**Comment by yoyo837** [comment]
Based on your log, For example:

<img width="895" alt="image" src="https://user-images.githubusercontent.com/6134547/155061914-dbdb090a-89d5-45e7-a738-e1dfcb323b9d.png">

**Comment by VindyaKaruturi** [comment]
> @sodatea we have updated the vue-cli version to 5.0.1 version for an existing project. This throws below error.
> 
> ValidationError: Invalid options object. Dev Server has been initialized using an options object that does not match the API schema. - options has an unknown property 'overlay'. These properties are valid: object { allowedHosts?, bonjour?, client?, compress?, devMiddleware?, headers?, historyApiFallback?, host?, hot?, http2?, https?, ipc?, liveReload?, magicHtml?, onAfterSetupMiddleware?, onBeforeSetupMiddleware?, onListening?, open?, port?, proxy?, server?, setupExitSignals?, setupMiddlewares?, static?, watchFiles?, webSocketServer? } ValidationError: Invalid options object. Dev Server has been initialized using an options object that does not match the API schema.
> 
> * options has an unknown property 'overlay'. These properties are valid:
>   object { allowedHosts?, bonjour?, client?, compress?, devMiddleware?, headers?, historyApiFallback?, host?, hot?, http2?, https?, ipc?, liveReload?, magicHtml?, onAfterSetupMiddleware?, onBeforeSetupMiddleware?, onListening?, open?, port?, proxy?, server?, setupExitSignals?, setupMiddlewares?, static?, watchFiles?, webSocketServer? }
>   at validate (C:\Users\o885500\Vista\azj-geolocation-ui\node_modules\webpack-dev-server\node_modules\schema-utils\dist\validate.js:115:11)
>   at new Server (C:\Users\o885500\Vista\azj-geolocation-ui\node_modules\webpack-dev-server\lib\Server.js:232:5)
>   at serve (C:\Users\o885500\Vista\azj-geolocation-ui\node_modules@vue\cli-service\lib\commands\serve.js:183:20)
>   at processTicksAndRejections (node:internal/process/task_queues:96:5)
>   error Command failed with exit code 1.



> But anyway, this new major version is stable and ready for use. Please enjoy.

**Comment by VindyaKaruturi** [comment]
> Based on your log, For example:
> 
> <img alt="image" width="895" src="https://user-images.githubusercontent.com/6134547/155061914-dbdb090a-89d5-45e7-a738-e1dfcb323b9d.png">

@yoyo837 Do we need to install webpack-dev-server? Because in our project we are not explicitly installed any webpack packages.

**Comment by Mister-Hope** [comment]
No, stop "in our project" why don't you check the lock file and see if you are actually using any webpack packages or not? Webpack is exactly which vue-cli is based on. 

The issue panel is not a place like stackoverflow. So that "Basic and simple questions" and "asking for help" is not welcomed here as you are actually noticing a lot of people on this tread. Try asking for help in vue forums or stackoverflow.

I DO think that it's ok for you to find problems and post here for the first time, but after someone already told you that this is not a issue and provide you the reson, please try to understand yourself. If you still don't get it, try to ask for help in [Discussion](https://github.com/vuejs/vue-cli/discussions)

For now, you are annoying me as one of the 59 people.

![image](https://user-images.githubusercontent.com/33315834/155062887-9ae83638-7a16-482d-8b3a-888065d81909.png)
